BUG: possible problem with np.add.accumulate on 32-bit windows

### Describe the issue:

In astropy, there are several failures with numpy 1.22. Most seem related to precision issues, perhaps due to the intel short vector math library. Those we can fix. A very strange one is a failure on win32 only: https://github.com/astropy/astropy/runs/4680891863?check_suite_focus=true, with the most relevant bit being:
```
>                   raw_field[1:, 1] = (np.add.accumulate(raw_field[:-1, 0]) *
                                        dtype.itemsize)
E                   TypeError: the resolved dtypes are not compatible with add.accumulate. Resolved (dtype('int32'), dtype('int32'), dtype('int32'))
```
The crucial bit seems to be a call that should be the equivalent of (sorry, don't have a windows machine so cannot test directly...),
```
np.add.accumulate(np.arange(10, dtype='>i4'))
```
On linux, this does not give any errors, though I can get the same `TypeError` by passing in an explicit dtype,
```
np.add.accumulate(np.arange(10, dtype='>i4'), dtype=np.int32)
```
This same line does not give an error on numpy 1.21.5, so that suggests this problem has to do with the new ufunc/dtype machinery, perhaps doing a wrong inference of the dtype on win32? Hence, cc @seberg.

### Reproduce the code example:

```python
import numpy as np
np.add.accumulate(np.arange(10, dtype='>i4'))
```


### Error message:

_No response_

### NumPy/Python version information:

platform win32 -- Python 3.9.9
3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)]
Numpy: 1.22.0

--- a/numpy/core/src/umath/ufunc_object.c
+++ b/numpy/core/src/umath/ufunc_object.c
@@ -2764,9 +2764,13 @@ reducelike_promote_and_resolve(PyUFuncObject *ufunc,
      * The first operand and output should be the same array, so they should
      * be identical.  The second argument can be different for reductions,
      * but is checked to be identical for accumulate and reduceat.
+     * Ideally, the type-resolver ensures that all are identical, but we do
+     * not enforce this here strictly.  Otherwise correct handling of
+     * byte-order changes (or metadata) requires a lot of care; see gh-20699.
      */
-    if (out_descrs[0] != out_descrs[2] || (
-            enforce_uniform_args && out_descrs[0] != out_descrs[1])) {
+    if (!PyArray_EquivTypes(out_descrs[0], out_descrs[2]) || (
+            enforce_uniform_args && !PyArray_EquivTypes(
+                    out_descrs[0], out_descrs[1]))) {
         PyErr_Format(PyExc_TypeError,
                 "the resolved dtypes are not compatible with %s.%s. "
                 "Resolved (%R, %R, %R)",

--- a/numpy/core/tests/test_ufunc.py
+++ b/numpy/core/tests/test_ufunc.py
@@ -2148,6 +2148,17 @@ class TestUfunc:
             # It would be safe, but not equiv casting:
             ufunc(a, c, out=out, casting="equiv")
 
+    def test_reducelike_byteorder_resolution(self):
+        # See gh-20699, byte-order changes need some extra care in the type
+        # resolution to make the following succeed:
+        arr_be = np.arange(10, dtype=">i8")
+        arr_le = np.arange(10, dtype="<i8")
+
+        assert np.add.reduce(arr_be) == np.add.reduce(arr_le)
+        assert_array_equal(np.add.accumulate(arr_be), np.add.accumulate(arr_le))
+        assert_array_equal(
+            np.add.reduceat(arr_be, [1]), np.add.reduceat(arr_le, [1]))
+
     def test_reducelike_out_promotes(self):
         # Check that the out argument to reductions is considered for
         # promotion.  See also gh-20455.

